src/Bicep.Core/TypeSystem/DeployTimeConstantVisitor.cs

```diff
@@ -0,0 +1,130 @@
+        {
+            this.model = model;
+            this.diagnosticWriter = diagnosticWriter;
+            this.runtimeValueAllowed = true;
+            this.currentSymbol = "";
+        }
+
+        // entry point for this visitor
+        public static void ValidateDeployTimeConstants(SemanticModel.SemanticModel model, IDiagnosticWriter diagnosticWriter)
+        {
+            new DeployTimeConstantVisitor(model, diagnosticWriter).Visit(model.Root.Syntax);
+        }
+
+        public override void VisitResourceDeclarationSyntax(ResourceDeclarationSyntax syntax)
+        {
+            this.runtimeValueAllowed = false;
+            base.VisitResourceDeclarationSyntax(syntax);
+            this.runtimeValueAllowed = true;
+        }
+
+        public override void VisitModuleDeclarationSyntax(ModuleDeclarationSyntax syntax)
+        {
+            this.runtimeValueAllowed = false;
+            base.VisitModuleDeclarationSyntax(syntax);
+            this.runtimeValueAllowed = true;
+        }
+
+        public override void VisitObjectPropertySyntax(ObjectPropertySyntax syntax)
+        {
+            bool priorRuntimeValueAllowed = this.runtimeValueAllowed;
+            // runtimeValueAllowed should only be false if we are going through a module or resource's objectPropertySyntax
+            if (!this.runtimeValueAllowed && syntax.Key is IdentifierSyntax keyIdentifier) 
+            {
+                this.currentSymbol = keyIdentifier.IdentifierName;
+                // right now we only check for the name property (using resourceNamePropertyName but the name is same for modules)
+                if (!keyIdentifier.IdentifierName.Equals(LanguageConstants.ResourceNamePropertyName, System.StringComparison.Ordinal))
+                {
+                    this.runtimeValueAllowed = true;
+                }
+            }
+            base.VisitObjectPropertySyntax(syntax);
+            // restore the value prior to the visit
+            this.runtimeValueAllowed = priorRuntimeValueAllowed;
+        }
+
+        public override void VisitPropertyAccessSyntax(PropertyAccessSyntax syntax)
+        {
+            if (!this.runtimeValueAllowed)
+            {
+                if (syntax.BaseExpression is VariableAccessSyntax variableAccessSyntax)
+                {
+                    // validate only on resource and module symbols
+                    // TODO these switch cases are so redundant.....
+                    switch (model.GetSymbolInfo(variableAccessSyntax))
+                    {
+                        case ResourceSymbol resourceSymbol:
+                            if ((resourceSymbol.Type is ResourceType resourceType) &&
+                            (resourceType.Body is ObjectType resourceBodyObj))
+                            {
+                                var property = syntax.PropertyName.IdentifierName;
+                                if (resourceBodyObj.Properties.TryGetValue(property, out var propertyType) &&
```

Comment:
>TryGetValue [](start = 63, length = 11)

We have some of this property lookup logic in `TypeAssignmentVisitor` and a slightly different version in `DeclaredTypeManager`. It's becoming a common enough operation that we should probably use shared logic. Otherwise, bugs like the discriminated object type comment above will slip in into the various forked implementation. It'd be great if you could commonize which ever existing implementation is appropriate. (It will also help with the ArrayAccess comment as well.)

Reply:
Couldn't figure out a feasible way to do this.